alf/algorithms/predictive_representation_learner.py

```diff
@@ -0,0 +1,360 @@
+        return self._target_field
+
+    def train_step(self, repr, state=()):
+        predicted_reward = self._decoder_net(repr)[0]
+        return AlgStep(
+            output=predicted_reward, state=state, info=predicted_reward)
+
+    def calc_loss(self, target, predicted, mask=None):
+        """Calculate the loss between ``target`` and ``predicted``.
+
+        Args:
+            target (Tensor): target to be predicted. Its shape is [T, B, ...]
+            predicted (Tensor): predicted target. Its shape is [T, B, ...]
+            mask (bool Tensor): indicating which target should be predicted.
+                Its shape is [T, B].
+        Returns:
+            LossInfo
+        """
+        loss = self._loss(predicted, target)
+        if self._debug_summaries and alf.summary.should_record_summaries():
+            with alf.summary.scope(self._name):
+
+                def _summarize1(pred, tgt, loss, mask, suffix):
+                    alf.summary.scalar(
+                        "explained_variance" + suffix,
+                        tensor_utils.explained_variance(pred, tgt, mask))
+                    safe_mean_hist_summary('predict' + suffix, pred, mask)
+                    safe_mean_hist_summary('target' + suffix, tgt, mask)
+                    safe_mean_summary("loss" + suffix, loss, mask)
+
+                def _summarize(pred, tgt, loss, mask, suffix):
+                    _summarize1(pred[0], tgt[0], loss[0], mask[0],
+                                suffix + "/current")
+                    if pred.shape[0] > 1:
+                        _summarize1(pred[1:], tgt[1:], loss[1:], mask[1:],
+                                    suffix + "/future")
+
+                if loss.ndim == 2:
+                    _summarize(predicted, target, loss, mask, '')
+                elif not self._summarize_each_dimension:
+                    m = mask
+                    if m is not None:
+                        m = m.unsqueeze(-1).expand_as(predicted)
+                    _summarize(predicted, target, loss, m, '')
+                else:
+                    for i in range(predicted.shape[2]):
+                        suffix = '/' + str(i)
+                        _summarize(predicted[..., i], target[..., i],
+                                   loss[..., i], mask, suffix)
+
+        if loss.ndim == 3:
+            loss = loss.mean(dim=2)
+
+        if mask is not None:
+            loss = loss * mask
+
+        return LossInfo(loss=loss * self._loss_weight)
+
+
+@gin.configurable
+class PredictiveRepresentationLearner(Algorithm):
+    """Learn representation based on the prediction of future values.
+
+    ``PredictiveRepresentationLearner`` contains 3 ``Network``s:
+    * encoding_net: it is used to encode the raw observation to a latent vector.
+    * dynamics_net: it is used to generate the future latent states from the
+        current latent state.
+    * decoder: it is used to decode the target values and calcuate the loss.
```

Comment:
Might be better to make it clear here that decoder is not purely a ``Network`` but an ``Algorithm``.

Reply:
updated